lib/liquid.lua

```diff
@@ -2547,28 +2549,39 @@ end
 -------------------------------------------------------------ParserContext end ---------------------------------------------------------
 -------------------------------------------------------------FileSystem begin ---------------------------------------------------------
 -- local FileSystem = {}
-function FileSystem:new( location )
+function FileSystem:new( get, error_handler )
     -- body
     local instance = {}
     setmetatable(instance, {__index = FileSystem})
-    instance.location = location
+    instance.get = get
     instance.text = nil
+    instance.error_handler = error_handler
     return instance
 end
 --
-function FileSystem:genertic_get( ... )
+function FileSystem:genertic_get( location )
+    local error_handler = self.error_handler
     -- body
     if self.get and type(self.get) == "function" then
-        local status, err = pcall(self.get, self.location)
-        if status then
-            return err
+        local ok, ret = pcall(self.get, location)
+
+        if not ret then
+            return error_handler(location, "cannot render empty template" )
+	      end
+
+        if ok then
+            return ret
         else
-            error("get template: " .. location .. " fail by user self defined get method" .. tostring(err))
+            return error_handler(location, "fail by user self defined get method: " .. tostring(ret))
```

Comment:
I think error codes are used a lot in OpenResty when functions `return ok, err` and `err` is some simple string.

I'd consider using tables for this and try to throw tables like https://github.com/3scale/apicast/commit/49a1ecfda9ec3dcc9b62af868995d5f6b8bcad03.
Then those tables can implement methods for equality or just compare metatables.

If you look at the current version there is just one string when there is empty template. IMO that is fine for now and can be handled just by comparing strings. I'd focus on this when there is some use case.